Public/Src/Cache/ContentStore/Distributed/NuCache/NuCacheCollectionUtilities.cs

```diff
@@ -237,29 +237,41 @@ public static IEnumerable<T> SkipOptimized<T>(this IReadOnlyList<T> original, in
         /// process, adding them to the same priority queue. This process repeats until the original enumerable has no
         /// more elements.
         /// </summary>
-        public static IEnumerable<T> QueryAndOrderInPages<T>(this IEnumerable<T> original, int pageSize, Comparer<T> comparer, Func<List<T>, IEnumerable<T>> query)
+        public static IEnumerable<T> QueryAndOrderInPages<T>(this IEnumerable<T> original, int pageSize, Comparer<T> comparer, Func<List<T>, IEnumerable<T>> query, float? takeoutFraction = null, int poolMultiplier = 2)
         {
+            Contract.Requires(pageSize > 0);
+            Contract.Requires(takeoutFraction == null || (takeoutFraction > 0 && takeoutFraction <= 1));
+            Contract.Requires(poolMultiplier > 0);
+
+            var poolSize = Math.Max(pageSize * poolMultiplier, pageSize);
+
+            // We either take the fraction given, or a single page at a time
+            var removalFraction = takeoutFraction ?? (1 / poolMultiplier);
+
             var source = original.GetEnumerator();
-            var queue = new PriorityQueue<T>(pageSize * 2, comparer);
+            var pool = new PriorityQueue<T>(poolSize, comparer);
             var sourceHasItems = true;
 
             while (true)
             {
-                if (sourceHasItems && queue.Count < pageSize)
+                if (sourceHasItems && pool.Count < poolSize)

```

Comment:
I am a confused about this if/else structure. As currently written, seems like
  while (true) {
    if (pool not full) {
      get another page
    } else {
      deliver results
    }
  }
Shouldn't it be
  while (true) {
    if (pool not full) {
      get another page
    }
    deliver results
  }
I don't want to fill the pool before we start delivering results.